File: src/checks.ts
import yaml from "js-yaml";
import { parse as parseToml } from "@iarna/toml";
import { join } from "path";
import { mkdir, writeFile, unlink } from "fs/promises";
import type { Format } from "./format.js";
import type { Reporter } from "./reporter.js";

export interface CheckOptions {
  preserve?: string[];
  checkCmd?: string;
  checkCmds?: Record<string, string>;
  reporter?: Reporter;
  label?: string;
  current?: number;
  total?: number;
}

export class ToolError extends Error {
  tool: string;
  constructor(tool: string, err: Error | string) {
    const message = typeof err === "string" ? err : (err.message ?? String(err));
    super(`${tool} tool failed: ${message}`);
    this.tool = tool;
  }
}

const DEFAULT_PRESERVE = ["code_blocks", "inline_code", "urls", "placeholders"];

const codeBlockRe = /```[\s\S]*?```/g;
const inlineCodeRe = /`[^`\n]+`/g;
const urlRe = /https?:\/\/[^\s)"'<>]+/g;
const placeholderRe = /\{[^\s{}]+\}/g;

export async function validate(
  root: string,
  format: Format,
  output: string,
  source: string,
  opts: CheckOptions,
): Promise<void> {
  if (opts.reporter && opts.label?.trim()) {
    opts.reporter.activity("Validating", opts.current ?? 0, opts.total ?? 0, opts.label);
  }

  // Syntax validation
  if (opts.reporter) {
    opts.reporter.tool("syntax-validator", "parse " + formatLabel(format));
  }
  const syntaxErr = validateSyntax(format, output);
  if (syntaxErr) {
    throw new ToolError("syntax-validator", syntaxErr);
  }

  // Preserve checks
  const preserveKinds = resolvePreserve(opts.preserve);
  if (Object.keys(preserveKinds).length > 0) {
    if (opts.reporter) {
      opts.reporter.tool("preserve-check", "verify preserved tokens");
    }
    const preserveErr = validatePreserve(output, source, preserveKinds);
    if (preserveErr) {
      throw new ToolError("preserve-check", preserveErr);
    }
  }

  // Custom command
  const cmd = selectCheckCmd(format, opts.checkCmd, opts.checkCmds);
  if (cmd) {
    if (opts.reporter) {
      opts.reporter.tool("custom-command", "run check_cmd");
    }
    await runExternal(root, cmd, output);
  }
}

function validateSyntax(format: Format, output: string): string | null {
  switch (format) {
    case "json":
      try {
        JSON.parse(output);
      } catch (err: any) {
        return `json invalid: ${err.message}`;
      }
      break;
    case "yaml":
      try {
        yaml.load(output);
      } catch (err: any) {
        return `yaml invalid: ${err.message}`;
      }
      break;
    case "po":
      return validatePO(output);
    case "markdown":
      return validateMarkdown(output);
  }
  return null;
}

function validateMarkdown(content: string): string | null {
  const lines = content.split("\n");
  if (lines.length === 0) return null;
  const first = lines[0].trim();
  if (first !== "---" && first !== "+++") return null;

  let end = -1;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i].trim() === first) {
      end = i;
      break;
    }
  }
  if (end === -1) {
    return `markdown frontmatter missing closing ${first}`;
  }

  const frontmatter = lines.slice(1, end).join("\n");
  if (first === "---") {
    try {
      yaml.load(frontmatter);
    } catch (err: any) {
      return `markdown frontmatter invalid yaml: ${err.message}`;
    }
    return null;
  }

  try {
    parseToml(frontmatter);
  } catch (err: any) {
    return `markdown frontmatter invalid toml: ${err.message}`;
  }
  return null;
}

function validatePO(content: string): string | null {
  const lines = content.split("\n");
  let state = "";
  let hasMsgid = false;
  let hasMsgstr = false;

  for (const rawLine of lines) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) continue;

    if (line.startsWith("msgid ")) {
      if (hasMsgid && !hasMsgstr) return "po entry missing msgstr";
      hasMsgid = true;
      hasMsgstr = false;
      state = "msgid";
      if (!hasQuotedString(line)) return "po msgid missing quoted string";
    } else if (line.startsWith("msgid_plural ")) {
      if (state !== "msgid") return "po msgid_plural without msgid";
      if (!hasQuotedString(line)) return "po msgid_plural missing quoted string";
    } else if (line.startsWith("msgstr")) {
      if (!hasMsgid) return "po msgstr without msgid";
      hasMsgstr = true;
      state = "msgstr";
      if (!hasQuotedString(line)) return "po msgstr missing quoted string";
    } else if (line.startsWith('"')) {
      if (!state) return "po stray quoted string";
    } else {
      return `po invalid line: ${line}`;
    }
  }
  if (hasMsgid && !hasMsgstr) return "po entry missing msgstr";
  return null;
}

function hasQuotedString(line: string): boolean {
  let count = 0;
  let escaped = false;
  for (const ch of line) {
    if (ch === "\\" && !escaped) {
      escaped = true;
      continue;
    }
    if (ch === '"' && !escaped) count++;
    escaped = false;
  }
  return count >= 2;
}

function resolvePreserve(preserve?: string[]): Record<string, boolean> {
  if (!preserve || preserve.length === 0) {
    return Object.fromEntries(DEFAULT_PRESERVE.map((k) => [k, true]));
  }
  for (const v of preserve) {
    if (v.trim().toLowerCase() === "none") return {};
  }
  return Object.fromEntries(preserve.map((k) => [k.trim().toLowerCase(), true]));
}

function extractPreservables(source: string, kinds: Record<string, boolean>): string[] {
  const tokens: string[] = [];
  const seen = new Set<string>();
  let text = source;

  if (kinds["code_blocks"]) {
    for (const match of text.matchAll(codeBlockRe)) {
      if (!seen.has(match[0])) {
        seen.add(match[0]);
        tokens.push(match[0]);
      }
    }
    text = text.replace(codeBlockRe, "");
  }
  if (kinds["inline_code"]) {
    for (const match of text.matchAll(inlineCodeRe)) {
      if (!seen.has(match[0])) {
        seen.add(match[0]);
        tokens.push(match[0]);
      }
    }
  }
  if (kinds["urls"]) {
    for (const match of text.matchAll(urlRe)) {
      if (!seen.has(match[0])) {
        seen.add(match[0]);
        tokens.push(match[0]);
      }
    }
  }
  if (kinds["placeholders"]) {
    for (const match of text.matchAll(placeholderRe)) {
      if (!seen.has(match[0])) {
        seen.add(match[0]);
        tokens.push(match[0]);
      }
    }
  }
  return tokens;
}

function validatePreserve(
  output: string,
  source: string,
  kinds: Record<string, boolean>,
): string | null {
  const preservables = extractPreservables(source, kinds);
  const missing: string[] = [];
  for (const token of preservables) {
    if (!output.includes(token)) {
      missing.push(token);
      if (missing.length >= 5) break;
    }
  }
  if (missing.length > 0) {
    return `preserved tokens missing from output: ${JSON.stringify(missing)}`;
  }
  return null;
}

function selectCheckCmd(format: Format, fallback?: string, cmds?: Record<string, string>): string {
  if (cmds) {
    const value = cmds[format];
    if (value?.trim()) return value;
  }
  return (fallback ?? "").trim();
}

function formatLabel(format: Format): string {
  switch (format) {
    case "json":
      return "JSON";
    case "yaml":
      return "YAML";
    case "po":
      return "PO";
    case "markdown":
      return "Markdown frontmatter";
    default:
      return format;
  }
}

async function runExternal(root: string, cmdTemplate: string, content: string): Promise<void> {
  if (!root) throw new Error("external check requires root path");

  const tmpDir = join(root, ".l10n", "tmp");
  await mkdir(tmpDir, { recursive: true });

  const tmpFile = join(tmpDir, `check-${Date.now()}-${Math.random().toString(36).slice(2)}.tmp`);
  await writeFile(tmpFile, content);

  try {
    const cmdText = cmdTemplate.replaceAll("{path}", tmpFile);
    const { execSync } = await import("child_process");
    try {
      execSync(cmdText, { cwd: root, stdio: "pipe" });
    } catch (err: any) {
      const output = (
        (err.stderr?.toString() ?? "") +
        "\n" +
        (err.stdout?.toString() ?? "")
      ).trim();
      throw new Error(`external check failed: ${err.message}\n${output}`);
    }
  } finally {
    await unlink(tmpFile).catch(() => {});
  }
}

File: internal/checks/checks.go
package checks

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/pelletier/go-toml/v2"
	"gopkg.in/yaml.v3"

	"github.com/tuist/l10n/internal/plan"
)

type Checker struct {
	Root string
}

type Options struct {
	Preserve  []string
	CheckCmd  string
	CheckCmds map[string]string
	Reporter  ToolReporter
}

var defaultPreserve = []string{"code_blocks", "inline_code", "urls", "placeholders"}

type ToolError struct {
	Tool string
	Err  error
}

func (e ToolError) Error() string {
	return fmt.Sprintf("%s tool failed: %v", e.Tool, e.Err)
}

func (e ToolError) Unwrap() error {
	return e.Err
}

type ToolReporter interface {
	Tool(name, detail string)
}

func (c Checker) Validate(ctx context.Context, format plan.Format, output string, source string, opts Options) error {
	if opts.Reporter != nil {
		opts.Reporter.Tool("syntax-validator", "parse "+formatLabel(format))
	}
	if err := validateSyntax(format, output); err != nil {
		return ToolError{Tool: "syntax-validator", Err: err}
	}

	preserveKinds := resolvePreserve(opts.Preserve)
	if len(preserveKinds) > 0 {
		if opts.Reporter != nil {
			opts.Reporter.Tool("preserve-check", "verify preserved tokens")
		}
		if err := validatePreserve(output, source, preserveKinds); err != nil {
			return ToolError{Tool: "preserve-check", Err: err}
		}
	}

	cmd := selectCheckCmd(format, opts.CheckCmd, opts.CheckCmds)
	if cmd != "" {
		if opts.Reporter != nil {
			opts.Reporter.Tool("custom-command", "run check_cmd")
		}
		if err := c.runExternal(ctx, cmd, output); err != nil {
			return ToolError{Tool: "custom-command", Err: err}
		}
	}

	return nil
}

func validateSyntax(format plan.Format, output string) error {
	switch format {
	case plan.FormatJSON:
		var v any
		if err := json.Unmarshal([]byte(output), &v); err != nil {
			return fmt.Errorf("json invalid: %w", err)
		}
	case plan.FormatYAML:
		var v any
		if err := yaml.Unmarshal([]byte(output), &v); err != nil {
			return fmt.Errorf("yaml invalid: %w", err)
		}
	case plan.FormatPO:
		if err := validatePO(output); err != nil {
			return err
		}
	case plan.FormatMarkdown:
		if err := validateMarkdown(output); err != nil {
			return err
		}
	}
	return nil
}

func resolvePreserve(preserve []string) map[string]bool {
	if len(preserve) == 0 {
		return sliceToSet(defaultPreserve)
	}
	for _, value := range preserve {
		if strings.TrimSpace(strings.ToLower(value)) == "none" {
			return map[string]bool{}
		}
	}
	return sliceToSet(preserve)
}

func sliceToSet(items []string) map[string]bool {
	set := map[string]bool{}
	for _, item := range items {
		item = strings.TrimSpace(strings.ToLower(item))
		if item == "" {
			continue
		}
		set[item] = true
	}
	return set
}

func validatePreserve(output, source string, kinds map[string]bool) error {
	preservables := extractPreservables(source, kinds)
	missing := []string{}
	for _, token := range preservables {
		if !strings.Contains(output, token) {
			missing = append(missing, token)
			if len(missing) >= 5 {
				break
			}
		}
	}
	if len(missing) > 0 {
		return fmt.Errorf("preserved tokens missing from output: %q", missing)
	}
	return nil
}

func extractPreservables(source string, kinds map[string]bool) []string {
	var tokens []string
	seen := map[string]bool{}
	text := source

	if kinds["code_blocks"] {
		for _, block := range codeBlockRe.FindAllString(text, -1) {
			if !seen[block] {
				seen[block] = true
				tokens = append(tokens, block)
			}
		}
		text = codeBlockRe.ReplaceAllString(text, "")
	}
	if kinds["inline_code"] {
		for _, code := range inlineCodeRe.FindAllString(text, -1) {
			if !seen[code] {
				seen[code] = true
				tokens = append(tokens, code)
			}
		}
	}
	if kinds["urls"] {
		for _, url := range urlRe.FindAllString(text, -1) {
			if !seen[url] {
				seen[url] = true
				tokens = append(tokens, url)
			}
		}
	}
	if kinds["placeholders"] {
		for _, ph := range placeholderRe.FindAllString(text, -1) {
			if !seen[ph] {
				seen[ph] = true
				tokens = append(tokens, ph)
			}
		}
	}

	return tokens
}

var (
	codeBlockRe   = regexp.MustCompile("(?s)```.*?```")
	inlineCodeRe  = regexp.MustCompile("`[^`\n]+`")
	urlRe         = regexp.MustCompile(`https?://[^\s)"'<>]+`)
	placeholderRe = regexp.MustCompile(`\{[^\s{}]+\}`)
)

func validateMarkdown(content string) error {
	lines := strings.Split(content, "\n")
	if len(lines) == 0 {
		return nil
	}
	first := strings.TrimSpace(lines[0])
	if first != "---" && first != "+++" {
		return nil
	}

	end := -1
	for i := 1; i < len(lines); i++ {
		if strings.TrimSpace(lines[i]) == first {
			end = i
			break
		}
	}
	if end == -1 {
		return fmt.Errorf("markdown frontmatter missing closing %s", first)
	}

	frontmatter := strings.Join(lines[1:end], "\n")
	if first == "---" {
		var v any
		if err := yaml.Unmarshal([]byte(frontmatter), &v); err != nil {
			return fmt.Errorf("markdown frontmatter invalid yaml: %w", err)
		}
		return nil
	}

	var v any
	if err := toml.Unmarshal([]byte(frontmatter), &v); err != nil {
		return fmt.Errorf("markdown frontmatter invalid toml: %w", err)
	}
	return nil
}

func validatePO(content string) error {
	scanner := bufio.NewScanner(strings.NewReader(content))
	state := ""
	hasMsgid := false
	hasMsgstr := false
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		switch {
		case strings.HasPrefix(line, "msgid "):
			if hasMsgid && !hasMsgstr {
				return errors.New("po entry missing msgstr")
			}
			hasMsgid = true
			hasMsgstr = false
			state = "msgid"
			if !hasQuotedString(line) {
				return errors.New("po msgid missing quoted string")
			}
		case strings.HasPrefix(line, "msgid_plural "):
			if state != "msgid" {
				return errors.New("po msgid_plural without msgid")
			}
			if !hasQuotedString(line) {
				return errors.New("po msgid_plural missing quoted string")
			}
		case strings.HasPrefix(line, "msgstr"):
			if !hasMsgid {
				return errors.New("po msgstr without msgid")
			}
			hasMsgstr = true
			state = "msgstr"
			if !hasQuotedString(line) {
				return errors.New("po msgstr missing quoted string")
			}
		case strings.HasPrefix(line, "\""):
			if state == "" {
				return errors.New("po stray quoted string")
			}
		default:
			return fmt.Errorf("po invalid line: %s", line)
		}
	}
	if err := scanner.Err(); err != nil {
		return err
	}
	if hasMsgid && !hasMsgstr {
		return errors.New("po entry missing msgstr")
	}
	return nil
}

func hasQuotedString(line string) bool {
	count := 0
	escaped := false
	for _, r := range line {
		if r == '\\' && !escaped {
			escaped = true
			continue
		}
		if r == '"' && !escaped {
			count++
		}
		escaped = false
	}
	return count >= 2
}

func formatLabel(format plan.Format) string {
	switch format {
	case plan.FormatJSON:
		return "JSON"
	case plan.FormatYAML:
		return "YAML"
	case plan.FormatPO:
		return "PO"
	case plan.FormatMarkdown:
		return "Markdown frontmatter"
	default:
		return string(format)
	}
}

func selectCheckCmd(format plan.Format, fallback string, cmds map[string]string) string {
	key := string(format)
	if cmds != nil {
		if value := cmds[key]; strings.TrimSpace(value) != "" {
			return value
		}
	}
	return strings.TrimSpace(fallback)
}

func (c Checker) runExternal(ctx context.Context, cmdTemplate string, content string) error {
	if c.Root == "" {
		return errors.New("external check requires root path")
	}
	tmpDir := filepath.Join(c.Root, ".l10n", "tmp")
	if err := os.MkdirAll(tmpDir, 0o755); err != nil {
		return err
	}
	tmpFile, err := os.CreateTemp(tmpDir, "check-*.tmp")
	if err != nil {
		return err
	}
	defer os.Remove(tmpFile.Name())
	if _, err := tmpFile.WriteString(content); err != nil {
		_ = tmpFile.Close()
		return err
	}
	if err := tmpFile.Close(); err != nil {
		return err
	}

	cmdText := strings.ReplaceAll(cmdTemplate, "{path}", tmpFile.Name())
	cmd := exec.CommandContext(ctx, "sh", "-c", cmdText)
	cmd.Dir = c.Root
	output, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("external check failed: %w\n%s", err, strings.TrimSpace(string(output)))
	}
	return nil
}
